lib/Monitor/Ultimagen/RunFolder.pm

```diff
@@ -0,0 +1,290 @@
+      $self->logcroak($error);
+    }
+  } else {
+    $self->info('Will create a new run for ' . $self->runfolder_path);
+    my $data = {
+      flowcell_id          => $self->ultima_runid,
+      folder_name          => $self->folder_name,
+      id_instrument        => $self->tracking_instrument()->id_instrument,
+      folder_path_glob     => $self->runfolder_glob,
+      team                 => 'SR',
+      id_instrument_format => $self->tracking_instrument()->id_instrument_format,
+      priority             => 1,
+      is_paired            => 0,
+      batch_id             => undef,
+    };
+
+    my $transaction = sub {
+      $run_row = $rs->create($data);
+      $self->info('Created run ' . $run_row->folder_name . ' with ID ' . $run_row->id_run);
+
+      my $runlane_rs = $run_row->result_source()->schema()->resultset('RunLane');
+      $runlane_rs->create({id_run => $run_row->id_run, position => 1});
+      $self->info("Created record for lane 1 of run_id " . $run_row->id_run);
+
+      $run_row->update_run_status(
+        'run in progress', $USERNAME, $self->date_created);
+      $self->_set_tags();
+    };
+    $rs->result_source()->schema()->txn_do($transaction);
+  }
+  return $run_row;
+}
+
+=head2 tracking_instrument
+
+Record representation of an instrument in the tracking database.
+
+The instrument record is retrieved uniquely (by DB definition).
+If no instrument is found, it exits with error.
+
+Returns a Result::Instrument instance from which instrument properties
+can be retrieved in the DB.
+
+=cut
+has q{tracking_instrument} => (
+  isa           => q{npg_tracking::Schema::Result::Instrument},
+  is            => q{ro},
+  lazy          => 1,
+  builder       => q{_build_tracking_instrument},
+);
+sub _build_tracking_instrument {
+  my $self = shift;
+  my $rs = $self->npg_tracking_schema->resultset('Instrument');
+  my $params = {
+    external_name => $INSTRUMENT_NAME
+  };
+  my @instrument_rows = $rs->search($params)->all();
+
+  my $instrument_count = scalar @instrument_rows;
+  if ($instrument_count == 0) {
+    $self->logcroak('No current instrument found in NPG tracking DB with name '
+      . $INSTRUMENT_NAME);
+  }
+
+  my $instrument_row = $instrument_rows[0];
+  $self->debug('Found instrument ' . $instrument_row->name());  
+  return $instrument_row;
+}
+
+sub _set_tags {
+  my ($self) = shift;
+  my @tags = (
+    'staging',
+    'multiplex'
+  );
+
+  foreach my $tag ( @tags ) {
+    $self->tracking_run()->set_tag($USERNAME, $tag);
+    $self->info("$tag tag is set");
+  }
+}
+
+=head2 process_run
+
+Core function of the class that is called on the run folder
+periodically to update dynamic properties of a run in the DB.
+If a run record does not exist, it is created.
+The tags and the following statuses are
+checked/assigned accordingly:
+- 'run in progress'   run basecalling is in progress
+- 'run mirrored'      run completed successfully
```

Comment:
The description here is 'all run data is copied from the instrument to teh staging run folder'. The run might have completed much earlier.